modulemd/modulemd-prioritizer.c

```diff
@@ -0,0 +1,209 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "modulemd.h"
+#include "modulemd-util.h"
+
+struct _ModulemdPrioritizer
+{
+  GObject parent_instance;
+
+  GHashTable *priorities;
+};
+
+G_DEFINE_TYPE (ModulemdPrioritizer, modulemd_prioritizer, G_TYPE_OBJECT)
+
+ModulemdPrioritizer *
+modulemd_prioritizer_new (void)
+{
+  return g_object_new (MODULEMD_TYPE_PRIORITIZER, NULL);
+}
+
+static void
+modulemd_prioritizer_finalize (GObject *object)
+{
+  ModulemdPrioritizer *self = (ModulemdPrioritizer *)object;
+
+  g_clear_pointer (&self->priorities, g_hash_table_unref);
+
+  G_OBJECT_CLASS (modulemd_prioritizer_parent_class)->finalize (object);
+}
+
+static void
+modulemd_prioritizer_class_init (ModulemdPrioritizerClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = modulemd_prioritizer_finalize;
+
+  /* This class has no public properties */
+  object_class->get_property = NULL;
+  object_class->set_property = NULL;
+}
+
+static void
+_modulemd_ptr_array_unref (gpointer ptr)
+{
+  g_ptr_array_unref ((GPtrArray *)ptr);
+}
+
+static void
+modulemd_prioritizer_init (ModulemdPrioritizer *self)
+{
+  self->priorities = g_hash_table_new_full (
+    g_int64_hash, g_int64_equal, g_free, _modulemd_ptr_array_unref);
+}
+
+
+/**
+ * modulemd_prioritizer_add:
+ * @objects: (array zero-terminated=1) (element-type GObject): A #GPtrArray of
+ * module-related objects loaded from a modulemd YAML stream.
+ * @priority: The priority of the YAML stream these were loaded from. Items at
+ * the same priority level will attempt to merge on conflict. Items at higher
+ * priority levels will replace on conflict.
+ *
+ * Returns: TRUE if the objects could be added without generating a conflict at
+ * this priority level. If a conflict was detected, this function returns FALSE
+ * and error is set. The internal state is undefined in the case of an error.
+ *
+ * Since: 1.3
+ */
+gboolean
+modulemd_prioritizer_add (ModulemdPrioritizer *self,
+                          GPtrArray *objects,
+                          gint64 priority,
+                          GError **error)
+{
+  GPtrArray *current_objects = NULL;
+  g_autoptr (GPtrArray) concat_objects = NULL;
+  g_autoptr (GPtrArray) merged_objects = NULL;
+  gint64 *prio = NULL;
+  gsize i;
+
+  g_return_val_if_fail (MODULEMD_IS_PRIORITIZER (self), FALSE);
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  prio = g_new0 (gint64, 1);
+  *prio = priority;
```

Comment:
why not simply `prio = &priotiry`?

Reply:
I originally had that and it took me almost two hours to discover that it was the source of a memory-corruption issue I was dealing with. The safest way to keep a key valid in GHashTable is to allocate it and associate that memory with the table.